pkg/webhook/tenant/validating.go

```diff
@@ -59,65 +58,71 @@ func Handler() capsulewebhook.Handler {
+func (h *handler) validateTenant(ctx context.Context, req admission.Request, clt client.Client, decoder *admission.Decoder) error {
+	tenant := &v1alpha1.Tenant{}
+	if err := decoder.Decode(req, tenant); err != nil {
+		return err
+	}
+
+	matched, _ := regexp.MatchString(`[a-z0-9]([-a-z0-9]*[a-z0-9])?`, tenant.GetName())
+	if !matched {
+		return fmt.Errorf("tenant name has forbidden characters")
+	}
+
+	// Validate ingressClasses regexp
+	if tenant.Spec.IngressClasses != nil && len(tenant.Spec.IngressClasses.Regex) > 0 {
+		if _, err := regexp.Compile(tenant.Spec.IngressClasses.Regex); err != nil {
+			return fmt.Errorf("unable to compile ingressClasses allowedRegex")
 		}
+	}
 
-		matched, _ := regexp.MatchString(`[a-z0-9]([-a-z0-9]*[a-z0-9])?`, tnt.GetName())
-		if !matched {
-			return admission.Denied("Tenant name has forbidden characters")
+	// Validate storageClasses regexp
+	if tenant.Spec.StorageClasses != nil && len(tenant.Spec.StorageClasses.Regex) > 0 {
+		if _, err := regexp.Compile(tenant.Spec.StorageClasses.Regex); err != nil {
+			return fmt.Errorf("unable to compile storageClasses allowedRegex")
 		}
+	}
 
-		// Validate ingressClasses regexp
-		if tnt.Spec.IngressClasses != nil && len(tnt.Spec.IngressClasses.Regex) > 0 {
-			if _, err := regexp.Compile(tnt.Spec.IngressClasses.Regex); err != nil {
-				return admission.Denied("Unable to compile ingressClasses allowedRegex")
-			}
+	// Validate containerRegistries regexp
+	if tenant.Spec.ContainerRegistries != nil && len(tenant.Spec.ContainerRegistries.Regex) > 0 {
+		if _, err := regexp.Compile(tenant.Spec.ContainerRegistries.Regex); err != nil {
+			return fmt.Errorf("unable to compile containerRegistries allowedRegex")
 		}
+	}
 
-		// Validate storageClasses regexp
-		if tnt.Spec.StorageClasses != nil && len(tnt.Spec.StorageClasses.Regex) > 0 {
-			if _, err := regexp.Compile(tnt.Spec.StorageClasses.Regex); err != nil {
-				return admission.Denied("Unable to compile storageClasses allowedRegex")
-			}
+	// Validate ingressHostnames regexp
+	if tenant.Spec.IngressHostnames != nil && len(tenant.Spec.IngressHostnames.Regex) > 0 {
+		if _, err := regexp.Compile(tenant.Spec.IngressHostnames.Regex); err != nil {
+			return fmt.Errorf("unable to compile ingressHostnames allowedRegex")
 		}
+	}
 
-		// Validate containerRegistries regexp
-		if tnt.Spec.ContainerRegistries != nil && len(tnt.Spec.ContainerRegistries.Regex) > 0 {
-			if _, err := regexp.Compile(tnt.Spec.ContainerRegistries.Regex); err != nil {
-				return admission.Denied("Unable to compile containerRegistries allowedRegex")
+	if tenant.Spec.IngressHostnames != nil && len(tenant.Spec.IngressHostnames.Exact) > 0 {
+		for _, h := range tenant.Spec.IngressHostnames.Exact {
+			tl := &v1alpha1.TenantList{}
+			if err := clt.List(ctx, tl, client.MatchingFieldsSelector{
+				Selector: fields.OneTermEqualSelector(".spec.ingressHostnames", h),
+			}); err != nil {
+				return err
+			}
+			if len(tl.Items) > 0 {
```

Comment:
@prometherion we should not raise error if the update is on the current tenant.

For example:

```
if len(tl.Items) > 0 && tl.Items[0].GetName() != tenant.GetName() {
    return fmt.Errorf("the allowed hostname %s is already used by the Tenant %s, cannot proceed", h, tl.Items[0].GetName())
}
```

Reply:
should be resolved by a2bc3a9